stablehlo/reference/Element.cpp

```diff
@@ -153,6 +186,39 @@ Element sine(const Element &e) {
   report_fatal_error(std::move(err));
 }
 
+Element tanh(const Element &e) {
+  Type type = e.getType();
+  if (isSupportedFloatType(type)) {
+    APFloat val = getFloatValue(e);
+    const llvm::fltSemantics &oldSemantics = val.getSemantics();
+    APFloat tanhVal(std::tanh(val.convertToDouble()));
+    bool roundingErr;
+    tanhVal.convert(oldSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type, FloatAttr::get(type, tanhVal));
+  } else if (isSupportedComplexType(type)) {
+    auto complex = getComplexValue(e);
+    Type elementType = type.cast<ComplexType>().getElementType();
+    const llvm::fltSemantics &oldSemantics = complex.real().getSemantics();
+    std::complex<double> complexVal(complex.real().convertToDouble(),
+                                    complex.imag().convertToDouble());
+    std::complex<double> tanhVal = std::tanh(complexVal);
+    bool roundingErr;
+    APFloat realAPF(tanhVal.real());
+    realAPF.convert(oldSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    APFloat imagAPF(tanhVal.imag());
+    imagAPF.convert(oldSemantics, APFloat::rmNearestTiesToEven, &roundingErr);
+    return Element(type,
+                   ArrayAttr::get(type.getContext(),
+                                  {FloatAttr::get(elementType, realAPF),
+                                   FloatAttr::get(elementType, imagAPF)}));
+  }
+
+  // Report error.
+  auto err = invalidArgument("Unsupported element type: %s",
+                             debugString(type).c_str());
+  report_fatal_error(std::move(err));
+}
```

Comment:
Can we avoid writing duplicate code for this? Let's find a way to share as much as possible between the implementations of sine/cosine/tanh in `Element.cpp`. One potential way of going about this would be to define a function with an interface along the lines of:
```
withUpcastToDouble(element,
  [](double value) { return std::tanh(value); },
  [](std::complex<double> value) { return std::tanh(value); })
```